### ISSUE TITLE
error: wrong version number (OpenSSL::SSL::SSLError)

### ISSUE THREAD

**Issue by langalex** [arkadiyt/ssrf_filter#42]
When trying to download certain URLs:

```ruby
SsrfFilter.get("https://images.cobot.me/uploads/plan/photo/aadce8899da1936c599bdcc49385c3b3/default_9622d00c-62ef-459c-9239-29d53ade828d.jpg")
```

The following error is displayed:

```
/Users/***/.asdf/installs/ruby/3.0.3/lib/ruby/gems/3.0.0/gems/net-protocol-0.1.3/lib/net/protocol.rb:46:in `connect_nonblock': SSL_connect returned=1 errno=0 state=error: wrong version number (OpenSSL::SSL::SSLError)
```

Not sure if this is the cause but the url above uses TLS 1.3.

I've narrowed it down to the [fetch_once](https://github.com/arkadiyt/ssrf_filter/blob/master/lib/ssrf_filter/ssrf_filter.rb#L172) method. Here is its code in a simplified version:

```ruby
url = "https://images.cobot.me/uploads/plan/photo/aadce8899da1936c599bdcc49385c3b3/default_9622d00c-62ef-459c-9239-29d53ade828d.jpg"
uri = URI.parse(url)

hostname = uri.hostname
ip_addresses = ::Resolv.getaddresses(hostname).map { |ip| ::IPAddr.new(ip) }

# culprit
uri.hostname = ip_addresses.sample.to_s

request = ::Net::HTTP::Get.new(uri)
request['host'] = "#{hostname}:443"

http_options = {}
http_options[:use_ssl] = (uri.scheme == 'https')

r = ::Net::HTTP.start(uri.hostname, uri.port, http_options) do |http|
  http.request(request)
end
```

Running the code above produces the error.  Commenting out the line under `# culprit` causes the error to go away.

**Comment by arkadiyt** [comment]
I'm not able to reproduce the issue:
```
irb(main):001:0> RUBY_VERSION
=> "3.0.3"
irb(main):002:0> SsrfFilter.get("https://images.cobot.me/uploads/plan/photo/aadce8899da1936c599bdcc49385c3b3/default_9622d00c-62ef-459c-9239-29d53ade
828d.jpg")
=> #<Net::HTTPOK 200 OK readbody=true>
```

Can you provide some more reproduction steps?

**Comment by langalex** [comment]
weird:

```ruby
irb(main):007:0> RUBY_VERSION
=> "3.0.3"
irb(main):008:0> SsrfFilter.get("https://images.cobot.me/uploads/plan/photo/aadce8899da1936c599bdcc49385c3b3/default_9622d00c-62ef-459c-
9239-29d53ade828d.jpg")
/Users/alexanderlang/.asdf/installs/ruby/3.0.3/lib/ruby/gems/3.0.0/gems/net-protocol-0.1.3/lib/net/protocol.rb:46:in `connect_nonblock': SSL_connect returned=1 errno=0 state=error: wrong version number (OpenSSL::SSL::SSLError)
```

will investigate further.

**Comment by arkadiyt** [comment]
Could you share your open ssl version?
```
irb(main):001:0> OpenSSL::VERSION
=> "2.2.1"
```
And also the output from this on your mac:
```
echo | openssl s_client -showcerts -connect images.cobot.me:443
```

**Comment by langalex** [comment]
```ruby
irb(main):001:0> OpenSSL::VERSION
=> "2.2.1"
```

```
$ echo | openssl s_client -showcerts -connect images.cobot.me:443

CONNECTED(00000006)
4370581036:error:1400410B:SSL routines:CONNECT_CR_SRVR_HELLO:wrong version number:/AppleInternal/Library/BuildRoots/b6051351-c030-11ec-96e9-3e7866fcf3a1/Library/Caches/com.apple.xbs/Sources/libressl/libressl-2.8/ssl/ssl_pkt.c:386:
---
no peer certificate available
---
No client certificate CA names sent
---
SSL handshake has read 5 bytes and written 0 bytes
---
New, (NONE), Cipher is (NONE)
Secure Renegotiation IS NOT supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
SSL-Session:
    Protocol  : TLSv1.2
    Cipher    : 0000
    Session-ID: 
    Session-ID-ctx: 
    Master-Key: 
    Start Time: 1653933021
    Timeout   : 7200 (sec)
    Verify return code: 0 (ok)
---
```

**Comment by arkadiyt** [comment]
Are you able to access that url normally, say from a browser even? Or curl?

That openssl output seems to indicate the server is not giving you back a TLS response at all. My output looks like this:

```
$ echo | openssl s_client -showcerts -connect images.cobot.me:443
CONNECTED(00000006)
depth=2 C = US, O = Amazon, CN = Amazon Root CA 1
verify return:1
depth=1 C = US, O = Amazon, OU = Server CA 1B, CN = Amazon
verify return:1
depth=0 CN = cdn.cobot.me
verify return:1
---
Certificate chain
 0 s:CN = cdn.cobot.me
   i:C = US, O = Amazon, OU = Server CA 1B, CN = Amazon
-----BEGIN CERTIFICATE-----
MIIGIjCCBQqgAwIBAgIQCsn5Zqxj7QUH815Ygnxi7zANBgkqhkiG9w0BAQsFADBG
MQswCQYDVQQGEwJVUzEPMA0GA1UEChMGQW1hem9uMRUwEwYDVQQLEwxTZXJ2ZXIg
Q0EgMUIxDzANBgNVBAMTBkFtYXpvbjAeFw0yMjAzMjQwMDAwMDBaFw0yMzA0MjIy
MzU5NTlaMBcxFTATBgNVBAMTDGNkbi5jb2JvdC5tZTCCASIwDQYJKoZIhvcNAQEB
BQADggEPADCCAQoCggEBAMSnCAaxkAjRlFzvZve6ejOD6911IBiEa0nIXrTvkno1
kJLFZrKaL0UL4XqgYixKPyU4qeOnE6mFJiAhpR5UcO5XbGS33quq52+cJrsRHqvv
vDIL4WedA6tgZv11PU81CueXKp1Yg3EXyQ5Qu1Z/3YJoC4SHjk8BdVwSbhmdHtuD
Ygp2EBvUABvfsRjpkfKeYGJHSIYfYWo54XjXyrZkj7m+BpFRkgkk1VaPLqFmVslf
LJE1HXwdb0K/iXopSivMFsXsQPJhqlZ07e5eowf8+3uS1dTn1ZK1ipewHjd/w48q
fOFxQXyDME1ZIgFgyr+Q2HfpQIJeS8Lkb3Jv3HtIJ+0CAwEAAaOCAzkwggM1MB8G
A1UdIwQYMBaAFFmkZgZSoHuVkjyjlAcnlnRb+T3QMB0GA1UdDgQWBBSXNpamtxSR
PMq3lFC63nb1xkx3PDBqBgNVHREEYzBhggxjZG4uY29ib3QubWWCDWNkbjEuY29i
b3QubWWCEGltYWdlczIuY29ib3QubWWCEGltYWdlczEuY29ib3QubWWCD2ltYWdl
cy5jb2JvdC5tZYINY2RuMi5jb2JvdC5tZTAOBgNVHQ8BAf8EBAMCBaAwHQYDVR0l
BBYwFAYIKwYBBQUHAwEGCCsGAQUFBwMCMD0GA1UdHwQ2MDQwMqAwoC6GLGh0dHA6
Ly9jcmwuc2NhMWIuYW1hem9udHJ1c3QuY29tL3NjYTFiLTEuY3JsMBMGA1UdIAQM
MAowCAYGZ4EMAQIBMHUGCCsGAQUFBwEBBGkwZzAtBggrBgEFBQcwAYYhaHR0cDov
L29jc3Auc2NhMWIuYW1hem9udHJ1c3QuY29tMDYGCCsGAQUFBzAChipodHRwOi8v
Y3J0LnNjYTFiLmFtYXpvbnRydXN0LmNvbS9zY2ExYi5jcnQwDAYDVR0TAQH/BAIw
ADCCAX0GCisGAQQB1nkCBAIEggFtBIIBaQFnAHUA6D7Q2j71BjUy51covIlryQPT
y9ERa+zraeF3fW0GvW4AAAF/uZjrcwAABAMARjBEAiB6VS1smS2TxOLlKkfv/Ycw
hWiUSPhaROe4rcN6TpRvNQIgDLVIHLt9YUSkrw+WyQmaG3Pz5VgUThIoS8N0kQij
XVYAdgA1zxkbv7FsV78PrUxtQsu7ticgJlHqP+Eq76gDwzvWTAAAAX+5mOuYAAAE
AwBHMEUCIGQ9WLniMq+xmpunckHF6p5rNmrjWzZb2d01Eoa1kSRkAiEAty4w/CTq
Dm8mo/Zn9FjajQJz7IcPo4uhwdfCr1fl7GMAdgCzc3cH4YRQ+GOG1gWp3BEJSnkt
sWcMC4fc8AMOeTalmgAAAX+5mOvLAAAEAwBHMEUCIQC/gH7aGEv1GEwA8NpdQvBW
ZNRt9cSz/gC7XR6iwdb9ygIgIIQzdayMYM1lTDWGbtyGXUaRy4WZyHGP0Hd0pvKu
QoUwDQYJKoZIhvcNAQELBQADggEBAG+cicJT/3NPCwIY/VpeWSn9GrRqMNhuEGQ9
qb8LkZRotHDaYzW8zJgYeMzp0w/KCBheZ5D7sp6kOYX8fo678r2WDYl5ho8Eml+Y
D2+G2KPFRZcumbsrfyaRGtaXw/+3X94shjLoQ7Y1oi80IZHCXpXQ1Xkcovj4+AvJ
5F3gOCe/tCCkMTHZdWzHOKnDnREaC40lwnIE6fB5/Nm38/xpahK/qhtY95RQoxtr
2gDQ7jYikelO5OPd8URnKXCpEUg/m00x+Y9LUDPYY10BPyvsAA5VwOFwXPJHhnMN
HT8oyD6Ba22tHISa3VeOgZ7nmD/W3j8VaS2JcQrD/RMcDQnyRfg=
-----END CERTIFICATE-----
 1 s:C = US, O = Amazon, OU = Server CA 1B, CN = Amazon
   i:C = US, O = Amazon, CN = Amazon Root CA 1
-----BEGIN CERTIFICATE-----
MIIESTCCAzGgAwIBAgITBn+UV4WH6Kx33rJTMlu8mYtWDTANBgkqhkiG9w0BAQsF
ADA5MQswCQYDVQQGEwJVUzEPMA0GA1UEChMGQW1hem9uMRkwFwYDVQQDExBBbWF6
b24gUm9vdCBDQSAxMB4XDTE1MTAyMjAwMDAwMFoXDTI1MTAxOTAwMDAwMFowRjEL
MAkGA1UEBhMCVVMxDzANBgNVBAoTBkFtYXpvbjEVMBMGA1UECxMMU2VydmVyIENB
IDFCMQ8wDQYDVQQDEwZBbWF6b24wggEiMA0GCSqGSIb3DQEBAQUAA4IBDwAwggEK
AoIBAQDCThZn3c68asg3Wuw6MLAd5tES6BIoSMzoKcG5blPVo+sDORrMd4f2AbnZ
cMzPa43j4wNxhplty6aUKk4T1qe9BOwKFjwK6zmxxLVYo7bHViXsPlJ6qOMpFge5
blDP+18x+B26A0piiQOuPkfyDyeR4xQghfj66Yo19V+emU3nazfvpFA+ROz6WoVm
B5x+F2pV8xeKNR7u6azDdU5YVX1TawprmxRC1+WsAYmz6qP+z8ArDITC2FMVy2fw
0IjKOtEXc/VfmtTFch5+AfGYMGMqqvJ6LcXiAhqG5TI+Dr0RtM88k+8XUBCeQ8IG
KuANaL7TiItKZYxK1MMuTJtV9IblAgMBAAGjggE7MIIBNzASBgNVHRMBAf8ECDAG
AQH/AgEAMA4GA1UdDwEB/wQEAwIBhjAdBgNVHQ4EFgQUWaRmBlKge5WSPKOUByeW
dFv5PdAwHwYDVR0jBBgwFoAUhBjMhTTsvAyUlC4IWZzHshBOCggwewYIKwYBBQUH
AQEEbzBtMC8GCCsGAQUFBzABhiNodHRwOi8vb2NzcC5yb290Y2ExLmFtYXpvbnRy
dXN0LmNvbTA6BggrBgEFBQcwAoYuaHR0cDovL2NydC5yb290Y2ExLmFtYXpvbnRy
dXN0LmNvbS9yb290Y2ExLmNlcjA/BgNVHR8EODA2MDSgMqAwhi5odHRwOi8vY3Js
LnJvb3RjYTEuYW1hem9udHJ1c3QuY29tL3Jvb3RjYTEuY3JsMBMGA1UdIAQMMAow
CAYGZ4EMAQIBMA0GCSqGSIb3DQEBCwUAA4IBAQCFkr41u3nPo4FCHOTjY3NTOVI1
59Gt/a6ZiqyJEi+752+a1U5y6iAwYfmXss2lJwJFqMp2PphKg5625kXg8kP2CN5t
6G7bMQcT8C8xDZNtYTd7WPD8UZiRKAJPBXa30/AbwuZe0GaFEQ8ugcYQgSn+IGBI
8/LwhBNTZTUVEWuCUUBVV18YtbAiPq3yXqMB48Oz+ctBWuZSkbvkNodPLamkB2g1
upRyzQ7qDn1X8nn8N8V7YJ6y68AtkHcNSRAnpTitxBKjtKPISLMVCx7i4hncxHZS
yLyKQXhw2W2Xs0qLeC1etA+jTGDK4UfLeC0SF7FSi8o5LL21L8IzApar2pR/
-----END CERTIFICATE-----
 2 s:C = US, O = Amazon, CN = Amazon Root CA 1
   i:C = US, ST = Arizona, L = Scottsdale, O = "Starfield Technologies, Inc.", CN = Starfield Services Root Certificate Authority - G2
-----BEGIN CERTIFICATE-----
MIIEkjCCA3qgAwIBAgITBn+USionzfP6wq4rAfkI7rnExjANBgkqhkiG9w0BAQsF
ADCBmDELMAkGA1UEBhMCVVMxEDAOBgNVBAgTB0FyaXpvbmExEzARBgNVBAcTClNj
b3R0c2RhbGUxJTAjBgNVBAoTHFN0YXJmaWVsZCBUZWNobm9sb2dpZXMsIEluYy4x
OzA5BgNVBAMTMlN0YXJmaWVsZCBTZXJ2aWNlcyBSb290IENlcnRpZmljYXRlIEF1
dGhvcml0eSAtIEcyMB4XDTE1MDUyNTEyMDAwMFoXDTM3MTIzMTAxMDAwMFowOTEL
MAkGA1UEBhMCVVMxDzANBgNVBAoTBkFtYXpvbjEZMBcGA1UEAxMQQW1hem9uIFJv
b3QgQ0EgMTCCASIwDQYJKoZIhvcNAQEBBQADggEPADCCAQoCggEBALJ4gHHKeNXj
ca9HgFB0fW7Y14h29Jlo91ghYPl0hAEvrAIthtOgQ3pOsqTQNroBvo3bSMgHFzZM
9O6II8c+6zf1tRn4SWiw3te5djgdYZ6k/oI2peVKVuRF4fn9tBb6dNqcmzU5L/qw
IFAGbHrQgLKm+a/sRxmPUDgH3KKHOVj4utWp+UhnMJbulHheb4mjUcAwhmahRWa6
VOujw5H5SNz/0egwLX0tdHA114gk957EWW67c4cX8jJGKLhD+rcdqsq08p8kDi1L
93FcXmn/6pUCyziKrlA4b9v7LWIbxcceVOF34GfID5yHI9Y/QCB/IIDEgEw+OyQm
jgSubJrIqg0CAwEAAaOCATEwggEtMA8GA1UdEwEB/wQFMAMBAf8wDgYDVR0PAQH/
BAQDAgGGMB0GA1UdDgQWBBSEGMyFNOy8DJSULghZnMeyEE4KCDAfBgNVHSMEGDAW
gBScXwDfqgHXMCs4iKK4bUqc8hGRgzB4BggrBgEFBQcBAQRsMGowLgYIKwYBBQUH
MAGGImh0dHA6Ly9vY3NwLnJvb3RnMi5hbWF6b250cnVzdC5jb20wOAYIKwYBBQUH
MAKGLGh0dHA6Ly9jcnQucm9vdGcyLmFtYXpvbnRydXN0LmNvbS9yb290ZzIuY2Vy
MD0GA1UdHwQ2MDQwMqAwoC6GLGh0dHA6Ly9jcmwucm9vdGcyLmFtYXpvbnRydXN0
LmNvbS9yb290ZzIuY3JsMBEGA1UdIAQKMAgwBgYEVR0gADANBgkqhkiG9w0BAQsF
AAOCAQEAYjdCXLwQtT6LLOkMm2xF4gcAevnFWAu5CIw+7bMlPLVvUOTNNWqnkzSW
MiGpSESrnO09tKpzbeR/FoCJbM8oAxiDR3mjEH4wW6w7sGDgd9QIpuEdfF7Au/ma
eyKdpwAJfqxGF4PcnCZXmTA5YpaP7dreqsXMGz7KQ2hsVxa81Q4gLv7/wmpdLqBK
bRRYh5TmOTFffHPLkIhqhBGWJ6bt2YFGpn6jcgAKUj6DiAdjd4lpFw85hdKrCEVN
0FE6/V1dN2RMfjCyVSRCnTawXZwXgWHxyvkQAiSr6w10kY17RSlQOYiypok1JR4U
akcjMS9cmvqtmg5iUaQqqcT5NJ0hGA==
-----END CERTIFICATE-----
 3 s:C = US, ST = Arizona, L = Scottsdale, O = "Starfield Technologies, Inc.", CN = Starfield Services Root Certificate Authority - G2
   i:C = US, O = "Starfield Technologies, Inc.", OU = Starfield Class 2 Certification Authority
-----BEGIN CERTIFICATE-----
MIIEdTCCA12gAwIBAgIJAKcOSkw0grd/MA0GCSqGSIb3DQEBCwUAMGgxCzAJBgNV
BAYTAlVTMSUwIwYDVQQKExxTdGFyZmllbGQgVGVjaG5vbG9naWVzLCBJbmMuMTIw
MAYDVQQLEylTdGFyZmllbGQgQ2xhc3MgMiBDZXJ0aWZpY2F0aW9uIEF1dGhvcml0
eTAeFw0wOTA5MDIwMDAwMDBaFw0zNDA2MjgxNzM5MTZaMIGYMQswCQYDVQQGEwJV
UzEQMA4GA1UECBMHQXJpem9uYTETMBEGA1UEBxMKU2NvdHRzZGFsZTElMCMGA1UE
ChMcU3RhcmZpZWxkIFRlY2hub2xvZ2llcywgSW5jLjE7MDkGA1UEAxMyU3RhcmZp
ZWxkIFNlcnZpY2VzIFJvb3QgQ2VydGlmaWNhdGUgQXV0aG9yaXR5IC0gRzIwggEi
MA0GCSqGSIb3DQEBAQUAA4IBDwAwggEKAoIBAQDVDDrEKvlO4vW+GZdfjohTsR8/
y8+fIBNtKTrID30892t2OGPZNmCom15cAICyL1l/9of5JUOG52kbUpqQ4XHj2C0N
Tm/2yEnZtvMaVq4rtnQU68/7JuMauh2WLmo7WJSJR1b/JaCTcFOD2oR0FMNnngRo
Ot+OQFodSk7PQ5E751bWAHDLUu57fa4657wx+UX2wmDPE1kCK4DMNEffud6QZW0C
zyyRpqbn3oUYSXxmTqM6bam17jQuug0DuDPfR+uxa40l2ZvOgdFFRjKWcIfeAg5J
Q4W2bHO7ZOphQazJ1FTfhy/HIrImzJ9ZVGif/L4qL8RVHHVAYBeFAlU5i38FAgMB
AAGjgfAwge0wDwYDVR0TAQH/BAUwAwEB/zAOBgNVHQ8BAf8EBAMCAYYwHQYDVR0O
BBYEFJxfAN+qAdcwKziIorhtSpzyEZGDMB8GA1UdIwQYMBaAFL9ft9HO3R+G9FtV
rNzXEMIOqYjnME8GCCsGAQUFBwEBBEMwQTAcBggrBgEFBQcwAYYQaHR0cDovL28u
c3MyLnVzLzAhBggrBgEFBQcwAoYVaHR0cDovL3guc3MyLnVzL3guY2VyMCYGA1Ud
HwQfMB0wG6AZoBeGFWh0dHA6Ly9zLnNzMi51cy9yLmNybDARBgNVHSAECjAIMAYG
BFUdIAAwDQYJKoZIhvcNAQELBQADggEBACMd44pXyn3pF3lM8R5V/cxTbj5HD9/G
VfKyBDbtgB9TxF00KGu+x1X8Z+rLP3+QsjPNG1gQggL4+C/1E2DUBc7xgQjB3ad1
l08YuW3e95ORCLp+QCztweq7dp4zBncdDQh/U90bZKuCJ/Fp1U1ervShw3WnWEQt
8jxwmKy6abaVd38PMV4s/KCHOkdp8Hlf9BRUpJVeEXgSYCfOn8J3/yNTd126/+pZ
59vPr5KW7ySaNRB6nJHGDn2Z9j8Z3/VyVOEVqQdZe4O/Ui5GjLIAZHYcSNPYeehu
VsyuLAOQ1xk4meTKCRlb/weWsKh/NEnfVqn3sF/tM+2MR7cwA130A4w=
-----END CERTIFICATE-----
---
Server certificate
subject=CN = cdn.cobot.me

issuer=C = US, O = Amazon, OU = Server CA 1B, CN = Amazon

---
No client certificate CA names sent
Peer signing digest: SHA256
Peer signature type: RSA-PSS
Server Temp Key: X25519, 253 bits
---
SSL handshake has read 5553 bytes and written 381 bytes
Verification: OK
---
New, TLSv1.3, Cipher is TLS_AES_128_GCM_SHA256
Server public key is 2048 bit
Secure Renegotiation IS NOT supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
Early data was not sent
Verify return code: 0 (ok)
---
DONE
```

**Comment by langalex** [comment]
yes, both browser and curl work.

**Comment by arkadiyt** [comment]
I'm not sure what is going on here then. The fact that you are not getting a certificate chain via `openssl s_client` makes me thing there's something going on with your environment

**Comment by langalex** [comment]
Yeah, at least it looks like it has nothing to do with ssrf_filter. Thanks for looking into it.

**Comment by arkadiyt** [comment]
@langalex someone else was having a similar issue [here](https://github.com/arkadiyt/ssrf_filter/issues/53) and I believe it's been resolved with the latest release (1.1.1) - could you take a look / confirm if that fixes it for you?

**Comment by langalex** [comment]
@arkadiyt can confirm the error is gone, thanks a lot 👍
